src/orange/OrangeTypes.cc

```diff
@@ -7,10 +7,72 @@
 //---------------------------------------------------------------------------//
 #include "OrangeTypes.hh"
 
+#include <type_traits>
+
+#include "corecel/Assert.hh"
 #include "corecel/io/EnumStringMapper.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/SoftEqual.hh"
 
 namespace celeritas
 {
+//---------------------------------------------------------------------------//
+/*!
+ * Use a relative error of \f$ \sqrt(\epsilon_\textrm{machine}) \f$ .
+ *
+ * Technically we're rounding the machine epsilon to a nearby power of 10. We
+ * could use numeric_limits<real_type>::epsilon instead.
+ */
+template<class T>
+Tolerance<T> Tolerance<T>::from_default(real_type length)
+{
+    constexpr real_type sqrt_emach = [] {
+        if constexpr (std::is_same_v<real_type, double>)
+        {
+            return 1.e-8;
+        }
+        else if constexpr (std::is_same_v<real_type, float>)
+        {
+            return 5.e-3f;
+        }
+    }();
+    static_assert(real_type{1} - ipow<2>(sqrt_emach) != real_type{1},
+                  "default tolerance is insufficient");
+
+    return Tolerance<T>::from_relative(sqrt_emach, length);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from the default "soft equivalence" tolerance.
+ */
+template<class T>
+Tolerance<T> Tolerance<T>::from_softequal()
+{
+    constexpr SoftEqual<> default_seq{};
+    return Tolerance<T>::from_relative(default_seq.rel(), default_seq.abs());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a relative tolerance and a length scale.
+ */
+template<class T>
+Tolerance<T> Tolerance<T>::from_relative(real_type rel, real_type length)
+{
+    CELER_VALIDATE(rel > 0 && rel < 1,
+                   << "tolerance " << rel
+                   << " is out of range [must be in (0,1)]");
+    CELER_VALIDATE(length > 0,
+                   << "length scale " << length
+                   << " is invalid [must be positive]");
+    Tolerance<T> result;
+    result.rel = rel;
+    result.abs = rel * length;
```

Comment:
Add a DBC  to ensure that the resulting absolute tolerance isn't below floating point epsilon

Reply:
I will put in a check that they're not denormalized.